numpyro/util.py

```diff
@@ -309,3 +311,263 @@ def soft_vmap(fn, xs, batch_ndims=1, chunk_size=None):
+        'node_data': node_data,
+        'edge_list': edge_list,
+    }
+
+
+def render_graph(graph_specification, render_distributions=False):
+    """
+    Create a graphviz object given a graph specification.
+
+    :param bool render_distributions: Show distribution of each RV in plot.
+    """
+    try:
+        import graphviz  # noqa: F401
+    except ImportError as e:
+        raise ImportError(
+            'Looks like you want to use graphviz (https://graphviz.org/) '
+            'to render your model. '
+            'You need to install `graphviz` to be able to use this feature. '
+            'It can be installed with `pip install graphviz`.'
+        ) from e
+
+    plate_groups = graph_specification['plate_groups']
+    plate_data = graph_specification['plate_data']
+    node_data = graph_specification['node_data']
+    edge_list = graph_specification['edge_list']
+
+    graph = graphviz.Digraph()
+
+    # add plates
+    plate_graph_dict = {
+        plate: graphviz.Digraph(name=f'cluster_{plate}')
+        for plate in plate_groups
+        if plate is not None
+    }
+    for plate, plate_graph in plate_graph_dict.items():
+        plate_graph.attr(label=plate.split('__CLONE')[0], labeljust='r', labelloc='b')
+
+    plate_graph_dict[None] = graph
+
+    # add nodes
+    for plate, rv_list in plate_groups.items():
+        cur_graph = plate_graph_dict[plate]
+
+        for rv in rv_list:
+            color = 'grey' if node_data[rv]['is_observed'] else 'white'
+            cur_graph.node(
+                rv, label=rv, shape='ellipse', style='filled', fillcolor=color
+            )
+
+    # add leaf nodes first
+    while len(plate_data) >= 1:
+        for plate, data in plate_data.items():
+            parent_plate = data['parent']
+            is_leaf = True
+
+            for plate2, data2 in plate_data.items():
+                if plate == data2['parent']:
+                    is_leaf = False
+                    break
+
+            if is_leaf:
+                plate_graph_dict[parent_plate].subgraph(plate_graph_dict[plate])
+                plate_data.pop(plate)
+                break
+
+    # add edges
+    for source, target in edge_list:
+        graph.edge(source, target)
+
+    # render distributions if requested
+    if render_distributions:
+        dist_label = ''
+        for rv, data in node_data.items():
+            rv_dist = data['distribution']
+            dist_label += f'{rv} ~ {rv_dist}\l'
+
+        graph.node(
+            'distribution_description_node', label=dist_label, shape='plaintext'
+        )
+
+    # return whole graph
+    return graph
+
+
+def render_model(
+    model,
+    model_args=None,
+    model_kwargs=None,
+    filename=None,
+    render_distributions=False,
+    num_tries=10,
+):
+    """
+    Wrap all functions needed to automatically render a model.
+
+    .. warning:: This utility does not support the :func:~numpyro.contrib.control_flow.scan primitive yet.
```

Comment:
nit:
```
:func:`~numpyro.contrib.control_flow.scan`
```

Reply:
Out of curiosity, how difficult would it be to also support rendering `scan`?

Reply:
To inspect the model and extract time information, it might be not so complicated. But it will be trickier to render. It is not clear to me how to represent the loops (which can have a long historical dependency: e.g. x_{t} depends on x_{t-1} and x_{t-2} in ARIMA) in graphics. Do you know which plate graphical notation is used to represent the loops? Probably we can start with history=1 and mimic some scan formats in [hmm enum example](http://num.pyro.ai/en/stable/examples/hmm_enum.html):
+ put all variables inside `scan` in a dashed-border rectangular (with label `t=0`)
+ create the same block with label `t=1` and put it side-to-side with block `t=0`
+ add three dots beside `t=1` block
+ add edges from `t=0` block to `t=1` block
+ add edges from outside-of-time-block nodes to inside-of-time-block nodes

Do you think it is feasible to automatically render such a graph? I can try to extract time information from the model if you want. Rendering the time plate would be truly neat, but it is complicated I guess. :)

Reply:
In the general case, I'd just render the whole chain without the `...` notation.
Or is it clear that the "sub-models" created by each scan iteration will always have the same structure?
Could a numpyro model potentially do something like this:
```python
def model():
    def foo(carry, i):
        # doesn't work, I know :-)
        if i == 6:
            numpyro.sample('special', dist.Beta(1, 1))
        else:
            numpyro.sample('default', dist.Beta(0.1, 0.1))

        return carry, i
        
    scan(foo, 0.01, jnp.arange(10))
```

Besides rendering dots, I also don't know of any better graphical representation.
Ignoring special spatial arrangements, we could probably treat `scan` as a "special plate" and surround it with a differently styled box.